### PR TITLE
SHOW-47: db: introduce schema versioning

### PR DESCRIPTION
**Pull request by markspolakovs**
Will be required for the next PR, splitting it out for ease of review

### CODE REVIEW

**Review by rmil — COMMENTED** [review]
Looks good! Definitely needed

It might be worhwhile adding a disclaimer to the README explaining the behaviour?

Also, how come you decided to go for an "in-house" migration tool, rather than something like [goose](https://github.com/pressly/goose)? Mainly curious here!

Also, apologies how the schema is stored is somewhat weird / novel

**Comment by rmil** [comment]
Also, if you could make a Jira ticket again for this, would be pog :3

**Comment by markspolakovs** [comment]
Hm, on second though, it looks like [embedding Goose](https://github.com/pressly/goose#embedded-sql-migrations) wouldn't actually be that hard. It would need need us to restructure the schema so that all migrations are in one place (rather than in each package (as far as I can tell), but part of me was wondering whether it's worth doing that anyway - in the current setup the order that each version's migrations are executed is non-deterministic anyway, which may be a problem if e.g. `mcr` depends on a table that's created by `livestream`.

**Comment by markspolakovs** [comment]
Have switched to Goose in a9cd975cfc212709cc53318233b58bbe42b10b09 - let me know your thoughts, happy to revert if you prefer the "in-house" approach.

File: db/migrations/migrations.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"database/sql"
+	"embed"
+	"fmt"
+	"math"
+
+	"github.com/pressly/goose/v3"
+)
+
+//go:embed *.sql
```
Comment (rmil) [inline db/migrations/migrations.go:12]: Does this work? Reason I ask is I've found go embedding to be cumbersome sometimes and needing to embed in the `cmd` directory (which is why I've stored the HTML templates in `cmd/main`

File: db/migrations/00001_initialise.sql [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    CONSTRAINT fk_livestream FOREIGN KEY (livestream_id) REFERENCES livestreams (livestream_id),
+    UNIQUE (integration_type, integration_id)
+);
+
+CREATE TABLE rtmp_outputs
+(
+    rtmp_output_id bigint GENERATED ALWAYS AS IDENTITY,
+    output_url     text NOT NULL,
+    PRIMARY KEY (rtmp_output_id)
+);
```
Comment (rmil) [inline db/migrations/00001_initialise.sql:32]: Oh wow I didn't realise these were not in their own schema, must be a left over from when I was using `sqlite`. I'm kinda tempted to move these to their own schema, but outside of scope of this PR!

**Review by rmil — APPROVED** [review]
Looks good! Would say add a little something to the README about when creating migrations refer to the goose docs?

File: db/migrations/migrations.go (hunk 1 above)
Comment (markspolakovs) [reply db/migrations/migrations.go:12]: It seemed to work when I tried it - it's relative to the `.go` file that has the `//go:embed` comment, so you could actually have the templates in another directory as long as it has a `.go` file with the embed export.

Comment (markspolakovs) [reply db/migrations/migrations.go:12]:
```sh
$ dropdb showtime && createdb showtime

$ go run cmd/init/init.go
2022/12/12 13:45:22 OK    00001_initialise.sql
2022/12/12 13:45:22 goose: no migrations to run. current version: 1
2022/12/12 13:45:22 successfully initialised showtime

```

(not sure what the "no migrations to run" line is all about, but it does definitely run them!)

Comment (rmil) [reply db/migrations/migrations.go:12]: Weird, makes it seem like it's got some sort of baseline and it's ignoring? But if it works then no issue!

You might be able to use the standalone goose tool to get a bit more insight into the state

Comment (markspolakovs) [reply db/migrations/migrations.go:12]: It seems to always print it when it's done:

https://github.com/pressly/goose/blob/b5826417d982a065ecbfa3d0d714ef2997a9775d/up.go#L112-L119

Comment (rmil) [reply db/migrations/migrations.go:12]: Ah, makes sense!

**Pull request merged**